index.js

```diff
@@ -1,22 +1,60 @@
-module.exports = Hook
-
 var register = require('./lib/register')
 var addHook = require('./lib/add')
 var removeHook = require('./lib/remove')
 
-function Hook () {
+var bind = Function.bind
+var bindable = bind.bind(bind)
+
+function bindApi (hook, state, name) {
+  const removeHookRef = bindable(removeHook, null).apply(null, name ? [state, null, name] : [state, null])
+  hook.api = { remove: removeHookRef }
+  hook.remove = removeHookRef
+
+  ;['before', 'error', 'after', 'wrap'].forEach(function (kind) {
+    const args = name ? [state, kind, name] : [state, kind]
+    hook[kind] = hook.api[kind] = bindable(addHook, null).apply(null, args)
+    hook.remove[kind] = hook.api.remove[kind] = bindable(removeHook, null).apply(null, args)
+  })
+}
+
+var singularHookState = {
+  registry: {}
+}
+var singularHookIterator = 0
+var singularHookDeprecationMessageDisplayed = false
+function HookSingular () {
+  if (!singularHookDeprecationMessageDisplayed) {
+    console.warn('[before-after-hook]: "Hook.Singular()" deprecation warning. Will be renamed to "Hook()" in the next major release.')
```

Comment:
I think I’d change the deprecation message to mention first that the `Hook` constructor is deprecated and that the user should use `Hook.Collection` instead. Then we could maybe add a https://git.io/ URL to the release notes which should have an explanation of the rename

Reply:
Well, this message is to indicate the `Hook.Singleton` is deprecated, so I'm not sure, but I think you're confusing the two, aren't you?
Actually, I think this deprecation message is a bit premature, why not show it when it's actually deprecated. What do you think?

Reply:
`Hook.Singular()` will only be deprecated with the next breaking version. Right now we will deprecate the usage of just `Hook`, so people use `Hook.Collection` if they want to keep the current functionality, which will keep working with the next breaking change when `Hook` will behave like `Hook.Singular` and `Hook.Singular` will be deprecated

Reply:
Yeah, if you correctly use the new API, you'll not get any warning, that's better indeed.